Sources/Crypto/Digests/BoringSSL/Digest_boring.swift

```diff
@@ -56,92 +56,134 @@ struct OpenSSLDigestImpl<H: BoringSSLBackedHashFunction> {
-            CCryptoBoringSSL_EVP_DigestFinal(self.contextPointer, digestPointer.baseAddress, &count)
+            return withUnsafeTemporaryAllocation(byteCount: digestSize, alignment: 1) { digestPointer in
+                assert(digestPointer.count == count)
+                CCryptoBoringSSL_EVP_DigestFinal_ex(&context, digestPointer.baseAddress, &count)
+                return body(UnsafeRawBufferPointer(digestPointer))
+            }
         }
-
-        return digestBytes
     }
 
     deinit {
-        CCryptoBoringSSL_EVP_MD_CTX_free(self.contextPointer)
+        self.withUnsafeMutablePointerToElements { elementsPtr in
+            let bytesToClear = self.digest.contextSize
+            memset_s(elementsPtr, bytesToClear, 0, bytesToClear)
+        }
+    }
+
+    private func withUnsafeContext<R>(_ body: (inout EVP_MD_CTX) -> R) -> R {
+        return self.withUnsafeMutablePointerToElements { elementsPtr in
+            return self.digest.withUnsafeContext(data: UnsafeMutableRawPointer(elementsPtr), body)
+        }
+    }
+
+    private func withUnsafeTemporaryContext<R>(_ body: (inout EVP_MD_CTX) -> R) -> R {
+        return self.withUnsafeMutablePointerToElements { elementsPtr in
+            return withUnsafeTemporaryAllocation(byteCount: self.digest.contextSize, alignment: 1) { temporaryBytes in
+                let source = UnsafeRawPointer(elementsPtr)
+                let destination = temporaryBytes.baseAddress!
+                destination.copyMemory(from: source, byteCount: self.digest.contextSize)
+                return self.digest.withUnsafeContext(data: destination, body)
+            }
+        }
     }
 }
 
 extension DigestContext {
     struct DigestType {
         var dispatchTable: OpaquePointer
+        var contextSize: Int
 
-        private init(_ dispatchTable: OpaquePointer) {
+        private init<T>(_ dispatchTable: OpaquePointer, _ type: T.Type) {
             self.dispatchTable = dispatchTable
+            self.contextSize = MemoryLayout<T>.size
         }
 
-        static let md5 = DigestType(CCryptoBoringSSL_EVP_md5())
+        static let md5 = DigestType(CCryptoBoringSSL_EVP_md5(), MD5_CTX.self)
 
-        static let sha1 = DigestType(CCryptoBoringSSL_EVP_sha1())
+        static let sha1 = DigestType(CCryptoBoringSSL_EVP_sha1(), SHA_CTX.self)
 
-        static let sha256 = DigestType(CCryptoBoringSSL_EVP_sha256())
+        static let sha256 = DigestType(CCryptoBoringSSL_EVP_sha256(), SHA256_CTX.self)
 
-        static let sha384 = DigestType(CCryptoBoringSSL_EVP_sha384())
+        static let sha384 = DigestType(CCryptoBoringSSL_EVP_sha384(), SHA384_CTX.self)
 
-        static let sha512 = DigestType(CCryptoBoringSSL_EVP_sha512())
+        static let sha512 = DigestType(CCryptoBoringSSL_EVP_sha512(), SHA512_CTX.self)
+
+        fileprivate typealias SHA384_CTX = SHA512_CTX
+
+        fileprivate func withUnsafeContext<R>(data: UnsafeMutableRawPointer, _ body: (inout EVP_MD_CTX) -> R) -> R {
```

Comment:
This code slightly awkwardly splits the baby: we store a digest specific context, and then rely on knowing how EVP_MD_CTX works internally to use it through the EVP_MD APIs, some of the time.

I don't think we should do it that way: we should either consistently use EVP_MD_CTX, or consistently use the typed digest, but not either-or.

If the goal is really to avoid the maximal number of allocations, that probably relies on us using the underlying digest contexts. That's fine, but if we're doing that we should be consistent with that, which means avoiding using the `EVP_MD_` APIs and instead using the `SHA_*` APIs. This will involve rebuilding the EVP_MD APIs, but that's not totally unreasonable if the goal is minimum number of allocations.

Reply:
Thanks for the feedback. I've rewritten the code to use the base `SHA_*` APIs.